File: frameworks/domain.core/src/main/java/ltd/highsoft/frameworks/domain/core/I18nMessage.java
package ltd.highsoft.frameworks.domain.core;

import lombok.*;
import org.apache.commons.lang3.StringUtils;

import java.util.List;

@ToString
@EqualsAndHashCode
public final class I18nMessage {

    public static final I18nMessage EMPTY = new I18nMessage("", new Object[]{});
    private final String code;
    private final Object[] data;

    public static I18nMessage message(String code, Object... data) {
        return new I18nMessage(code, data);
    }

    public String format(MessageResolver messageResolver) {
        if (StringUtils.isBlank(code)) return StringUtils.EMPTY;
        return messageResolver.resolve(code, data);
    }

    public String code() {
        return code;
    }

    public List<Object> data() {
        return List.of(data);
    }

    private I18nMessage(String code, Object[] data) {
        this.code = code;
        this.data = data;
    }

}

File: frontends/hare.webui/src/commons/i18n/I18nMessage.ts
import {culture} from 'commons/i18n/CultureAccessor';
import {Message} from 'commons/i18n/Message';

export class I18nMessage implements Message {
  public readonly code: string;
  private readonly args: unknown[];

  public constructor(code: string, ...args: unknown[]) {
    this.code = code;
    this.args = args;
  }

  public toString(): string {
    return culture.localize(this.code, this.args);
  }
}
